src/librbd/Group.cc

```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+	<< cpp_strerror(-r) << dendl;
+      return r;
+    }
+    cls_snaps.insert(cls_snaps.end(), snaps_page.begin(), snaps_page.end());
+
+  } while (r == max_read);
+
+  return 0;
+}
+
+int group_snap_remove(librados::IoCtx& group_ioctx, const char *group_name,
+		      const char *snap_name)
+{
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  librados::Rados rados(group_ioctx);
+
+  std::vector<cls::rbd::GroupSnapshot> snaps;
+  std::vector<C_SaferCond*> on_finishes;
+  int ret_code = 0;
+
+  string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+				 group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading consistency group id object: "
+	       << cpp_strerror(r)
+	       << dendl;
+    return r;
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  r = group_snap_list_cls(group_ioctx, group_name, snaps);
+  if (r < 0) {
+    return r;
+  }
+  cls::rbd::GroupSnapshot *gs = nullptr;
+  int n = snaps.size();
+  for (int i = 0; i < n; ++i) {
+    if (snaps[i].name == string(snap_name)) {
+      gs = &snaps[i];
+    }
+  }
+  if (gs == nullptr) {
+    return -ENOENT;
+  }
+
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+
+  std::string image_snap_name;
+
+  ldout(cct, 20) << "Removing snapshots" << dendl;
+  n = gs->snaps.size();
+
+  for (int i = 0; i < n; ++i) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
+    r = rados.ioctx_create2(gs->snaps[i].pool, *image_io_ctx);
+    if (r < 0) {
+      ldout(cct, 1) << "Failed to create io context for image" << dendl;
+    }
+
+    std::string image_name;
+    r = cls_client::dir_get_name(image_io_ctx, RBD_DIRECTORY,
+			     gs->snaps[i].image_id, &image_name);
+    if (r < 0) {
+      goto cleanup;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx(image_name, "", "",
+					       *image_io_ctx, false);
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    image_ctx->state->open(on_finish);
+
+    ictxs.push_back(image_ctx);
+    on_finishes.push_back(on_finish);
+  }
+
+  ret_code = 0;
+  for (int i = 0; i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0) {
+      delete ictxs[i];
+      ictxs[i] = nullptr;
+      ret_code = r;
+    }
+    if (ictxs[i]->id != gs->snaps[i].image_id) {
+      ldout(cct, 1) << "An image was renamed during snapshot removal. Try again later." << dendl;
+      ret_code = -EAGAIN;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+
+  ldout(cct, 20) << "Opened participating images. Deleting snapshots themselves." << dendl;
+
+  image_snap_name = calc_ind_image_snap_name(group_ioctx.get_id(),
+					     group_id,
+					     gs->id);
+  for (int i = 0; i < n; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    ldout(cct, 20) << "Removing individual snapshot with name: " << image_snap_name << dendl;
+    int r = ictx->state->refresh_if_required();
```

Comment:
Just opened the image -- no need to refresh (plus it will be handled automatically within ```Operations``` if required)